importer/wdi_importer.py

```diff
@@ -1,63 +1,922 @@
+    os.makedirs(wdi_downloads_save_location)
+else:
+    shutil.rmtree(wdi_downloads_save_location)
+    os.makedirs(wdi_downloads_save_location)
+
+# create a directory for holding the backup sql files
+if not os.path.exists(wdi_backups_location):
+    os.makedirs(wdi_backups_location)
+
+logger = logging.getLogger('importer')
+start_time = time.time()
+
+logger.info("Getting the zip file")
+request_header = {'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.95 Safari/537.36'}
+r = requests.get(wdi_zip_file_url, stream=True, headers=request_header)
+if r.ok:
+    with open(wdi_downloads_save_location + 'wdi.zip', 'wb') as out_file:
+        shutil.copyfileobj(r.raw, out_file)
+    logger.info("Saved the zip file to disk.")
+    z = zipfile.ZipFile(wdi_downloads_save_location + 'wdi.zip')
+    excel_filename = wdi_downloads_save_location + z.namelist()[0]  # there should be only one file inside the zipfile, so we will load that one
+    z.extractall(wdi_downloads_save_location)
+    r = None  # we do not need the request anymore
+    logger.info("Successfully extracted the zip file")
+else:
+    logger.error("The file could not be downloaded. Stopping the script...")
+    sys.exit("Could not download file.")
+
+wdi_category_name_in_db = 'World Development Indicators'  # set the name of the root category of all data that will be imported by this script
+
+
+import_history = ImportHistory.objects.filter(import_type='wdi')
+
+
+# if wdi imports were never performed
+if not import_history:
+    logger.info("This is the very first WDI data import.")
+    backup_sql_file = open(os.path.join(wdi_backups_location, '%s.sql' %
+                           timezone.now().strftime("%Y-%m-%d %H:%M:%S")), 'w', encoding='utf-8')
+
+    wb = load_workbook(excel_filename, read_only=True)
+
+    series_ws = wb['Series']
+    data_ws = wb['Data']
+    country_ws = wb['Country']
+
+    column_number = 0  # this will be reset to 0 on each new row
+    row_number = 0   # this will be reset to 0 if we switch to another worksheet, or start reading the worksheet from the beginning one more time
+
+    global_cat = {}  # global catalog of indicators
+
+    # data in the worksheets is not loaded into memory at once, that causes RAM to quickly fill up
+    # instead, we go through each row and cell one-by-one, looking at each piece of data separately
+    # this has the disadvantage of needing to traverse the worksheet several times, if we need to look up some rows/cells again
+
+    for row in series_ws.rows:
+        row_number += 1
+        indicator = None
+        for cell in row:
+            if row_number > 1:
+                column_number += 1
+                if column_number == 1:
+                    indicator = cell.value
+                    global_cat[cell.value] = {}
+                if column_number == 2:
+                    global_cat[indicator]['category'] = cell.value.split(':')[0]
```

Comment:
We can probably make this code faster and less verbose by assigning global_cat[indicator] to a variable on column_number == 1 instead of doing a lookup for every cell